src/stronghold/db.rs

```diff
@@ -3,112 +3,92 @@
-    }
-}
-
 #[async_trait]
 impl DatabaseProvider for StrongholdAdapter {
     async fn get(&mut self, k: &[u8]) -> Result<Option<Vec<u8>>> {
-        // Lazy load the snapshot (if the path is set).
-        if self.snapshot_path.is_some() {
-            self.read_stronghold_snapshot().await?;
-        }
-
-        let location = location_from_key(k);
-        let (data, status) = self.stronghold.lock().await.read_from_store(location).await;
-
-        // XXX: this theoretically indicates a non-existent key, but what about other errors?
-        if let ResultMessage::Error(err) = status {
-            debug!("Stronghold reported \"{}\", but we treat it as \"key not found\".", err);
-            return Ok(None);
-        }
-
-        let locked_key = self.key.lock().await;
-        let key = if let Some(key) = &*locked_key {
-            key
+        let data = match self
+            .stronghold
+            .lock()
+            .await
+            .get_client(PRIVATE_DATA_CLIENT_PATH)?
+            .store()
+            .get(k)?
+        {
+            Some(data) => data,
+            None => return Ok(None),
+        };
+
+        let locked_key_provider = self.key_provider.lock().await;
+        let key_provider = if let Some(key_provider) = &*locked_key_provider {
+            key_provider
         } else {
             return Err(Error::StrongholdKeyCleared);
         };
+        let buffer = key_provider.try_unlock()?;
+        let buffer_ref = buffer.borrow();
 
-        decrypt(&data, key).map(Some)
+        decrypt(&data, buffer_ref.deref()).map(Some)
     }
 
     async fn insert(&mut self, k: &[u8], v: &[u8]) -> Result<Option<Vec<u8>>> {
-        // Lazy load the snapshot (if the path is set).
-        if self.snapshot_path.is_some() {
-            self.read_stronghold_snapshot().await?;
-        }
-
-        let old_value = self.get(k).await?;
-
         let encrypted_value = {
-            let locked_key = self.key.lock().await;
-            let key = if let Some(key) = &*locked_key {
-                key
+            let locked_key_provider = self.key_provider.lock().await;
+            let key_provider = if let Some(key_provider) = &*locked_key_provider {
+                key_provider
             } else {
                 return Err(Error::StrongholdKeyCleared);
             };
+            let buffer = key_provider.try_unlock()?;
+            let buffer_ref = buffer.borrow();
 
-            encrypt(v, key)?
+            encrypt(v, buffer_ref.deref())?
         };
 
-        let location = location_from_key(k);
-        let status = self
+        Ok(self
             .stronghold
             .lock()
             .await
-            .write_to_store(location, encrypted_value, None)
-            .await;
-
-        if let ResultMessage::Error(err) = status {
-            return Err(Error::StrongholdProcedureError(err));
-        }
-
-        Ok(old_value)
+            .get_client(PRIVATE_DATA_CLIENT_PATH)?
```

Comment:
I wonder, if this would be easier to actually have each object dereferenced explicitly.